src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -62,5 +63,68 @@ public abstract class PipeReader
         /// Cancel the pending <see cref="ReadAsync"/> operation. If there is none, cancels next <see cref="ReadAsync"/> operation, without completing the <see cref="PipeWriter"/>.
         /// </summary>
         public abstract void OnWriterCompleted(Action<Exception, object> callback, object state);
+
+        /// <summary>
+        /// Asynchronously reads the bytes from the <see cref="PipeReader"/> and writes them to the specified stream, using a specified buffer size and cancellation token.
+        /// </summary>
+        /// <param name="destination">The stream to which the contents of the current stream will be copied.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                SequencePosition consumed = default;
+
+                try
+                {
+                    ReadResult result = await ReadAsync(cancellationToken).ConfigureAwait(false);
+                    ReadOnlySequence<byte> buffer = result.Buffer;
+
+                    consumed = buffer.Start;
+
+                    if (result.IsCanceled)
+                    {
+                        throw new OperationCanceledException();
+                    }
```

Comment:
Maybe:
```C#
if (result.IsCanceled)
{
    cancellationToken.ThrowIfCancellationRequested();
    throw new OperationCanceledException();
}
```
so that if the cancellation was triggered by the token, the exception will contain it?

Reply:
Kinda edgy. That means you passed a token and called CancelPendingRead()

Reply:
Oh, maybe I misunderstood something... if the token has cancellation requested and the ReadAsync observes it, does it throw or does it return a result with IsCanceled == true?

Reply:
> Oh, maybe I misunderstood something... if the token has cancellation requested and the ReadAsync observes it, does it throw or does it return a result with IsCanceled == true?

It throws. IsCanceled is only true if you call CancelPendingRead

Reply:
Ok, then what you have makes sense.